lib/icinga/service-apply.cpp

```diff
@@ -121,9 +121,14 @@ bool Service::EvaluateApplyRule(const Host::Ptr& host, const ApplyRule& rule)
 
 void Service::EvaluateApplyRules(const Host::Ptr& host)
 {
-	for (ApplyRule& rule : ApplyRule::GetRules("Service")) {
-		CONTEXT("Evaluating 'apply' rules for host '" + host->GetName() + "'");
+	CONTEXT("Evaluating 'apply' rules for host '" + host->GetName() + "'");
 
+	for (auto& rule : ApplyRule::GetRules(Service::TypeInstance, Host::TypeInstance)) {
+		if (EvaluateApplyRule(host, rule))
+			rule.AddMatch();
+	}
+
+	for (auto& rule : ApplyRule::GetRules(Service::TypeInstance, nullptr)) {
 		if (EvaluateApplyRule(host, rule))
 			rule.AddMatch();
 	}
```

Comment:
I'd rather normalize the input data to only have `Host::TypeInstance` as a key in the map rather than have this duplicated here for this special case.

Reply:
Too bad you don't have lib/icinga/ types in lib/config/ (AFAIK).

Reply:
That normalization could also happen when it's still a string value. For example in `config_parser.yy` or `ApplyRule::AddRule()`.

Reply:
That would be like if source is service and target is "", make target host. That would run for every apply rule of all types (and btw. malloc() for all actually affected ones). I prefer my version.

Reply:
Your other PR already shows how this leads to weird special casing down the road:
* https://github.com/Icinga/icinga2/blob/ad39b77d328ef387a159e3deab777055bdaacb1b/lib/config/applyrule-targeted.cpp#L57
* https://github.com/Icinga/icinga2/blob/ad39b77d328ef387a159e3deab777055bdaacb1b/lib/icinga/service-apply.cpp#L126

> That would be like if source is service and target is ""

No, that shouldn't require hard-coding of any of this but rather something like

```c++
if (target == "") {
    target = ApplyRule::GetTargetTypes(type)[0];
}
```

around there:

https://github.com/Icinga/icinga2/blob/f2563cc8904e73dada16d390f5a93c3b89a56f5b/lib/config/config_parser.yy#L1158-L1185